src/utils/wallet.js

```diff
@@ -77,30 +81,120 @@ const wallet = {
    * @memberof Wallet
    * @inner
    */
-  executeGenerateWallet(words, passphrase, pin, password, loadHistory) {
-    if (loadHistory) {
+  async startWallet(words, passphrase, pin, password, routerHistory, fromXpriv = false) {
+    // Set loading addresses screen to show
+    store.dispatch(loadingAddresses(true));
+    // When we start a wallet from the locked screen, we need to unlock it in the storage
+    oldWalletUtil.unlock();
+
+    // This check is important to set the correct network on storage and redux
+    const data = await version.checkApiVersion();
+
+    // Before cleaning loaded data we must save in redux what we have of tokens in localStorage
+    const dataToken = tokens.getTokens();
+    store.dispatch(reloadData({ tokens: dataToken }));
+
+    // If we've lost redux data, we could not properly stop the wallet object
+    // then we don't know if we've cleaned up the wallet data in the storage
+    // If it's fromXpriv, then we can't clean access data because we need that
+    oldWalletUtil.cleanLoadedData({ cleanAccessData: !fromXpriv});
+
+    const connection = new Connection({
+      network: data.network,
+      servers: [helpers.getServerURL()],
+    });
+
+    const beforeReloadCallback = () => {
       store.dispatch(loadingAddresses(true));
-      const fullPromise = new Promise((resolve, reject) => {
-        const versionPromise = version.checkApiVersion();
-        versionPromise.then(() => {
-          // Load history from address
-          const promise = hathorLib.wallet.executeGenerateWallet(words, passphrase, pin, password, loadHistory);
-          promise.then(() => {
-            this.afterLoadAddressHistory();
-            resolve();
-          });
-        });
-      });
-      return fullPromise;
-    } else {
-      return hathorLib.wallet.executeGenerateWallet(words, passphrase, pin, password, loadHistory);
+    };
+
+    let xpriv = null;
+    if (fromXpriv) {
+      xpriv = oldWalletUtil.getXprivKey(pin);
     }
+
+    const walletConfig = {
+      seed: words,
+      xpriv,
+      store: STORE,
+      passphrase,
+      connection,
+      beforeReloadCallback,
+    };
+
+    const wallet = new HathorWallet(walletConfig);
+
+    store.dispatch(setWallet(wallet));
+
+    const serverInfo = await wallet.start({ pinCode: pin, password });
+    // TODO should we save server info?
+    //store.dispatch(setServerInfo(serverInfo));
+    wallet.on('state', (state) => {
+      if (state === HathorWallet.ERROR) {
+        // ERROR
+        // TODO Should we show an error screen and try to restart the wallet?
+      } else if (state === HathorWallet.READY) {
+        // READY
+        const historyTransactions = wallet.getTxHistory();
+        store.dispatch(loadWalletSuccess(historyTransactions));
+      }
+    });
+
+    wallet.on('new-tx', (tx) => {
+      let message = '';
+      if (helpers.isBlock(tx)) {
```

Comment:
Let's open another PR to refactor this to another file? I feel we should have an internal notification bus just like we have in the full-node. It might be useful for future features.

I'm suggesting another PR because it feels too much for this one.

Reply:
Created the issue: https://github.com/HathorNetwork/hathor-wallet/issues/194